pep-9999.rst

```diff
@@ -0,0 +1,464 @@
+Sponsor: TBD
+Discussions-To: TBD
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 27-Feb-2025
+Python-Version: 3.12
+Post-History: 
+Resolution:
+
+
+Abstract
+========
+
+This PEP proposes adding a mechanism for deep immutability to
+Python via a new builtin function ``freeze(obj)``. This function
+takes a reference to an object and recursively render the object
+and all objects it references immutable. (This is *deep*
+immutability --- just making the first object immutable is called
+*shallow* immutability.)
+
+Deep immutability will provide stronger guarantees against
+unintended modifications, improving correctness, security, and
+parallel execution safety.
+
+In many respects, this proposal can be thought of as an
+extension of `PEP 683 --- Immortal Objects <https://peps.python.org/pep-0683/>`_
+most importantly including run-time checks that guarantee that
+an immutable object does not change, and with the key difference
+that it permits reference counts to change! In this PEP, we rely
+on the GIL to ensure the correctness of reference counts of 
+immutable objects, but we have several planned memory management 
+extensions.
+
+Immutability in action:
+
+.. code-block:: python
+
+	class Foo: pass
+
+	f = Foo()
+	g = Foo()
+	h = Foo()
+
+	f.f = g
+	g.f = h
+	h.f = g # cycle!
+
+	g.x = "African Swallow" # OK
+	freeze(f) # Makes, f, g and h immutable
+	g.x = "European Swallow" # Throws immutability exception
+
+
+Motivation
+==========
+
+
+Ensuring Data Integrity
+-----------------------
+
+Python programs frequently manipulate large, interconnected data
+structures such as dictionaries, lists, and user-defined objects.
+Unintentional mutations can introduce subtle and
+difficult-to-debug errors. By allowing developers to explicitly
+freeze objects and their transitive dependencies, Python can
+provide stronger correctness guarantees for data processing
+pipelines, functional programming paradigms, and API boundaries
+where immutability is beneficial.
+
+
+Eliminating Data Races in Concurrent Code
+-----------------------------------------
+
+Python’s Global Interpreter Lock (GIL) mitigates many data race
+issues, but as Python evolves towards improved multi-threading and
+parallel execution (e.g., subinterpreters and free-threaded Python
+efforts), data races on shared mutable objects become a more
+pressing concern. A deep immutability mechanism ensures that
+shared objects are not modified concurrently, enabling safer
+multi-threaded and parallel computation.
+
+See also the discussion about extensions further down in this
+document.
+
+
+Optimisations and Caching Benefits
+----------------------------------
+
+Immutable objects provide opportunities for optimisation, such as
+structural sharing, memoization, and just-in-time (JIT)
+compilation techniques (specialising for immutable data, e.g.
+fixed shape, fewer barriers, inlining, etc.). Freezing objects can
```

Comment:
There is already this to a degree due to `tuple `and `frozenset` and the basic types like `int` and `str` being treated as immutable for use in caching and hashing. Open question whether immutability can integrate with that system, but we are not doing that at present and probably would want that to be proposed future work.

Reply:
Good point. I'm namedropping these now.